src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -0,0 +1,885 @@
+    // - closes the websocket connection and the output thread
+    void AzureConnection::Close()
+    {
+        if (!_connected)
+        {
+            return;
+        }
+
+        if (!_closing.exchange(true))
+        {
+            _canProceed.notify_all();
+            if (_state == State::termConnected)
+            {
+                // Close the websocket connection
+                auto closedTask = _cloudShellSocket.close();
+                closedTask.wait();
+            }
+
+            // Tear down our output thread
+            WaitForSingleObject(_hOutputThread.get(), INFINITE);
+            _hOutputThread.reset();
+        }
+    }
+
+    DWORD WINAPI AzureConnection::StaticOutputThreadProc(LPVOID lpParameter)
+    {
+        AzureConnection* const pInstance = (AzureConnection*)lpParameter;
+        return pInstance->_OutputThread();
+    }
+
+    // Method description:
+    // - this is the output thread, where we initiate the connection to Azure
+    //  and establish a websocket connection
+    // Return value:
+    // - return status
+    DWORD AzureConnection::_OutputThread()
+    {
+        while (true)
+        {
+            try
+            {
+                switch (_state)
+                {
+                // Initial state, check if the user has any stored connection settings and allow them to login with those
+                // or allow them to login with a different account or allow them to remove the saved settings
+                case State::accessStored:
+                {
+                    RETURN_IF_FAILED(_accessHelper());
+                    break;
+                }
+                // User has no saved connection settings or has opted to login with a different account
+                // Azure authentication happens here
+                case State::deviceFlow:
+                {
+                    RETURN_IF_FAILED(_deviceFlowHelper());
+                    break;
+                }
+                // User has multiple tenants in their Azure account, they need to choose which one to connect to
+                case State::tenantChoice:
+                {
+                    RETURN_IF_FAILED(_tenantChoiceHelper());
+                    break;
+                }
+                // Ask the user if they want to save these connection settings for future logins
+                case State::storeTokens:
+                {
+                    RETURN_IF_FAILED(_storeHelper());
+                    break;
+                }
+                // Connect to Azure, we only get here once we have everything we need (tenantID, accessToken, refreshToken)
+                case State::termConnecting:
+                {
+                    RETURN_IF_FAILED(_connectHelper());
+                    break;
+                }
+                // We are connected, continuously read from the websocket until its closed
+                case State::termConnected:
+                {
+                    while (true)

```

Comment:
Why didn't this state get it's own helper method too?

Reply:
It felt like the real meat of the output thread (since its the state we'll be in _most_ of the time) so it really was just a stylistic choice to leave it there. Can change it if anyone feels strongly about it